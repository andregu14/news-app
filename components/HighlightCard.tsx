import { View, Text, ViewProps } from "./Themed";
import {
  StyleSheet,
  useColorScheme,
  useWindowDimensions,
  Pressable,
} from "react-native";
import Colors from "@/constants/Colors";
import { Image } from "expo-image";

type HighlightCardProps = ViewProps & {
  description?: string;
  image?: string;
  onPress?: () => void;
  testID?: string;
  imageTestID?: string;
};

const randomImage = "https://picsum.photos/170/200";

export default function HighlightCard({
  description,
  image,
  style,
  onPress,
  testID,
  imageTestID,
  ...otherProps
}: HighlightCardProps) {
  const colorScheme = useColorScheme() ?? "light";
  const themeColors = Colors[colorScheme];
  const { width } = useWindowDimensions();

  const cardWidth = Math.min(width * 0.4, 340);

  return (
    <Pressable
      onPress={onPress}
      style={({ pressed }) => [
        styles.container,
        {
          width: cardWidth,
          borderColor: themeColors.borderColor,
          opacity: pressed ? 0.8 : 1,
          backgroundColor: pressed
            ? colorScheme === "dark"
              ? "#333"
              : "#f5f5f5"
            : "transparent",
        },
      ]}
      testID={testID}
    >
      <View
        style={[
          {
            width: cardWidth,
            backgroundColor: "transparent",
          },
          style,
        ]}
        {...otherProps}
      >
        <Image
          style={styles.imageContainer}
          source={{ uri: image ?? randomImage }}
          contentFit="cover"
          testID={imageTestID}
        />
        <Text
          ellipsizeMode={"tail"}
          numberOfLines={4}
          style={styles.descriptionText}
        >
          {description
            ? description
            : "Lorem ipsum dolor sit amet Lorem ipsum dolor sit amet"}
        </Text>
      </View>
    </Pressable>
  );
}

const styles = StyleSheet.create({
  container: {
    height: 200,
    borderRadius: 14,
    alignItems: "center",
    marginHorizontal: 7.5,
    borderWidth: 1,
    overflow: "hidden",
  },
  imageContainer: {
    width: "100%",
    height: "50%",
  },
  descriptionText: {
    padding: 10,
  },
});
