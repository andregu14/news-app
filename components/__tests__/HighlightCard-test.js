import { render, fireEvent, screen, configure } from '@testing-library/react-native';
import HighlightCard from '../HighlightCard';

const mockUseColorScheme = jest.fn(() => 'light');
const mockUseWindowDimensions = jest.fn(() => ({
    width: 320,
    height: 640,
    scale: 1,
    fontScale: 1
}));

jest.mock('react-native/Libraries/Utilities/useColorScheme', () => ({
    default: mockUseColorScheme
}));

jest.mock('react-native/Libraries/Utilities/useWindowDimensions', () => ({
    default: mockUseWindowDimensions
}));

beforeEach(() => {
    mockUseColorScheme.mockReturnValue('light');
    mockUseWindowDimensions.mockReturnValue({
        width: 320,
        height: 640,
        scale: 1,
        fontScale: 1
    });
});

describe('HighlightCard', () => {
    it('renders correctly with description and image props and matches snapshot', () => {
        const testDescription = "Test Description";
        const testImage = "https://example.com/test-image.png";

        const component = render(
            <HighlightCard
                department="tecnologia"
                description={testDescription}
                image={testImage}
                testID="highlight-card"
                imageTestID="highlight-card-image"
            />
        );

        expect(screen.getByText(testDescription)).toBeTruthy();
        expect(screen.getByAccessibilityHint("Toque para abrir o artigo completo")).toBeTruthy();

        const snapshot = component.toJSON();
        expect(snapshot).toMatchSnapshot();

        const snapshotString = JSON.stringify(snapshot);
        expect(snapshotString).toContain('highlight-card-image');
        expect(snapshotString).toContain(testImage);
    });

    it('calls onPress prop when pressed', () => {
        const onPressMock = jest.fn();

        render(
            <HighlightCard
                department="tecnologia"
                onPress={onPressMock}
                testID="highlight-card"
            />
        );

        const cardElement = screen.getByTestId('highlight-card');
        fireEvent.press(cardElement);

        expect(onPressMock).toHaveBeenCalledTimes(1);
    });
});

describe('Accessibility', () => {
    it('has proper accessibility properties', () => {
        render(
            <HighlightCard
                department="tecnologia"
                onPress={jest.fn()}
                testID="highlight-card"
            />
        );

        const card = screen.getByTestId('highlight-card');
        expect(card.props.accessible).toBe(true);
        expect(card.props.accessibilityRole).toBe('button');
    });

    it('provides meaningful content descriptions', () => {
        const customDescription = "Artigo sobre tecnologia";

        render(
            <HighlightCard
                department="tecnologia"
                description={customDescription}
                onPress={jest.fn()}
            />
        );

        expect(screen.getByText(customDescription)).toBeTruthy();
        expect(screen.getByAccessibilityHint("Toque para abrir o artigo completo")).toBeTruthy();
    });
});
