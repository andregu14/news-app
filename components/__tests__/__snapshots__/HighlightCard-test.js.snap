// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HighlightCard renders correctly with description and image props and matches snapshot 1`] = `
<View
  accessibilityHint="Toque para abrir o artigo completo"
  accessibilityLabel="tecnologia: Test Description"
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  hitSlop={
    {
      "bottom": 8,
      "left": 8,
      "right": 8,
      "top": 8,
    }
  }
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "borderRadius": 14,
        "borderWidth": 1,
        "height": 200,
        "marginHorizontal": 7.5,
        "overflow": "hidden",
      },
      {
        "backgroundColor": "#fff",
        "borderColor": "#eee",
        "opacity": 1,
        "width": 128,
      },
    ]
  }
  testID="highlight-card"
>
  <View
    style={
      [
        {
          "backgroundColor": "#fff",
        },
        [
          {
            "backgroundColor": "transparent",
            "width": 128,
          },
          undefined,
        ],
      ]
    }
  >
    <ViewManagerAdapter_ExpoImage
      accessibilityElementsHidden={true}
      accessible={false}
      containerViewRef={"[React.ref]"}
      contentFit="cover"
      contentPosition={
        {
          "left": "50%",
          "top": "50%",
        }
      }
      height="50%"
      nativeViewRef={"[React.ref]"}
      onError={[Function]}
      onLoad={[Function]}
      onLoadStart={[Function]}
      onProgress={[Function]}
      placeholder={[]}
      source={
        [
          {
            "uri": "https://example.com/test-image.png",
          },
        ]
      }
      style={
        {
          "height": "50%",
          "width": "100%",
        }
      }
      testID="highlight-card-image"
      transition={null}
      width="100%"
    />
    <View
      accessible={false}
      importantForAccessibility="no-hide-descendants"
      style={
        [
          {
            "backgroundColor": "#fff",
          },
          [
            {
              "borderRadius": 20,
              "paddingHorizontal": 12,
              "paddingVertical": 6,
              "position": "absolute",
              "right": 5,
              "top": 5,
            },
            {
              "backgroundColor": "#000",
            },
          ],
        ]
      }
    >
      <Text
        accessible={false}
        style={
          [
            {
              "color": "#000",
            },
            [
              {
                "fontSize": 12,
                "fontWeight": 500,
              },
              {
                "color": "#fff",
              },
            ],
          ]
        }
      >
        tecnologia
      </Text>
    </View>
    <Text
      ellipsizeMode="tail"
      numberOfLines={4}
      style={
        [
          {
            "color": "#000",
          },
          {
            "padding": 10,
          },
        ]
      }
    >
      Test Description
    </Text>
    <View
      style={
        [
          {
            "backgroundColor": "#fff",
          },
          [
            {
              "alignSelf": "center",
              "borderRadius": 20,
              "bottom": -15,
              "height": 10,
              "position": "absolute",
              "width": "90%",
            },
            {
              "backgroundColor": "#fff",
            },
          ],
        ]
      }
    />
  </View>
</View>
`;
